src/components/Frame/Frame.tsx

```diff
@@ -66,340 +56,292 @@ const APP_FRAME_NAV = 'AppFrameNav';
-            className={navClassName}
-            onKeyDown={this.handleNavKeydown}
-            id={APP_FRAME_NAV}
-            key="NavContent"
-            hidden={mobileNavHidden}
-          >
-            {navigation}
-            <button
-              type="button"
-              className={styles.NavigationDismiss}
-              onClick={this.handleNavigationDismiss}
-              aria-hidden={
-                mobileNavHidden ||
-                (!isNavigationCollapsed && !showMobileNavigation)
-              }
-              aria-label={intl.translate(
-                'Polaris.Frame.Navigation.closeMobileNavigationLabel',
-              )}
-              tabIndex={tabIndex}
-            >
-              <Icon source={MobileCancelMajorMonotone} color="white" />
-            </button>
-          </div>
-        </CSSTransition>
-      </TrapFocus>
-    ) : null;
+  const showToast = useCallback((toast: ToastPropsWithID) => {
+    setToastMessages((toastMessages) => {
+      const hasToastById = Boolean(
+        toastMessages.find(({id}) => id === toast.id),
+      );
 
-    const loadingMarkup =
-      loadingStack > 0 ? (
-        <div className={styles.LoadingBar} id={APP_FRAME_LOADING_BAR}>
-          <Loading />
-        </div>
-      ) : null;
+      return hasToastById ? toastMessages : [...toastMessages, toast];
+    });
+  }, []);
 
-    const contextualSaveBarMarkup = (
-      <CSSAnimation
-        in={showContextualSaveBar}
-        className={styles.ContextualSaveBar}
-        type={AnimationType.Fade}
-      >
-        <ContextualSaveBar {...this.contextualSaveBar} />
-      </CSSAnimation>
+  const hideToast = useCallback(({id}: ToastID) => {
+    setToastMessages((toastMessages) =>
+      toastMessages.filter(({id: toastId}) => id !== toastId),
     );
-
-    const topBarMarkup = topBar ? (
-      <div
-        className={styles.TopBar}
-        {...layer.props}
-        {...dataPolarisTopBar.props}
-        id={APP_FRAME_TOP_BAR}
+  }, []);
+
+  const startLoading = useCallback(() => {
+    setLoadingStack((loadingStack) => loadingStack + 1);
+  }, []);
+
+  const stopLoading = useCallback(() => {
+    setLoadingStack((loadingStack) => Math.max(0, loadingStack - 1));
+  }, []);
+
+  const setContextualSaveBar = useCallback(
+    (props: ContextualSaveBarProps) => {
+      setContextualSaveBarProps({...props});
+      forceTrueShowContextualSaveBar();
+    },
+    [forceTrueShowContextualSaveBar],
+  );
+
+  const removeContextualSaveBar = useCallback(() => {
+    setContextualSaveBarProps(null);
+    forceFalseShowContextualSaveBar();
+  }, [forceFalseShowContextualSaveBar]);
+
+  useEffect(() => {
+    if (globalRibbon && globalRibbonContainer.current) {
+      const globalRibbonHeight = globalRibbonContainer.current.offsetHeight;
+      setRootProperty(
+        GLOBAL_RIBBON_CUSTOM_PROPERTY,
+        `${globalRibbonHeight}px`,
+        null,
+      );
+    }
+  }, [globalRibbon]);
+
+  const context = useMemo(
```

Comment:
I was initially kinda confused why this exists and why we couldn't define a variable instead of dealing with useMemo, but it's because the this object gets set as the FrameContext and we don't want to trigger the context changing every time we render the frame. Nice catch!